### PR TITLE
refactor(api): use hl id directly in nvim_buf_set_extmark

### PR DESCRIPTION
**Pull request by glepnir**
make  `*_hl_group`  only support `nil|int|string` and convert to hl id directly.

### CODE REVIEW

File: src/nvim/api/extmark.c [hunk 1]
```diff
@@ -540,36 +541,29 @@ Integer nvim_buf_set_extmark(Buffer buffer, Integer ns_id, Integer line, Integer
...
-  } hls[] = {
-    { "hl_group"           , &opts->hl_group           , &hl.hl_id            },
-    { "sign_hl_group"      , &opts->sign_hl_group      , &sign.hl_id       },
-    { "number_hl_group"    , &opts->number_hl_group    , &sign.number_hl_id     },
-    { "line_hl_group"      , &opts->line_hl_group      , &sign.line_hl_id       },
-    { "cursorline_hl_group", &opts->cursorline_hl_group, &sign.cursorline_hl_id },
-    { NULL, NULL, NULL },
-  };
+  if (opts->sign_hl_group) {
+    sign.hl_id = (int)opts->sign_hl_group;
```
Comment (bfredl) [inline src/nvim/api/extmark.c:550]: don't need these assignments be in if-statements. then we could just use `opts->sign_hl_group || opts->cursorline_hl_group || ...` in the end for the `sign.flags` check.

File: src/nvim/lua/converter.c [hunk 2]
```diff
@@ -1324,7 +1325,27 @@ void nlua_pop_keydict(lua_State *L, void *retval, FieldHashfn hashy, char **err_
...
+        if (lua_type(L, -1) == LUA_TSTRING) {
+          String name = nlua_pop_String(L, err);
+          int hl_id = 0;
+          if (!ERROR_SET(err) && name.size > 0) {
+            hl_id = syn_check_group(name.data, name.size);
+          }
+          api_free_string(name);
+          *(Integer *)mem = hl_id;
+        } else if (lua_type(L, -1) == LUA_TNUMBER) {
+          *(Integer *)mem = nlua_pop_Integer(L, err);
```
Comment (bfredl) [inline src/nvim/lua/converter.c:1338]: we can just do this in an `else`, after we checked for String, `nlua_pop_Integer` will do the rest.

File: src/nvim/lua/converter.c [hunk 3]
```diff
@@ -1324,7 +1325,24 @@ void nlua_pop_keydict(lua_State *L, void *retval, FieldHashfn hashy, char **err_
     if (field->type == kObjectTypeNil) {
       *(Object *)mem = nlua_pop_Object(L, true, err);
     } else if (field->type == kObjectTypeInteger) {
-      *(Integer *)mem = nlua_pop_Integer(L, err);
+      if (field->is_hlgroup) {
+        if (lua_type(L, -1) == LUA_TSTRING) {
+          String name = nlua_pop_String(L, err);
```
Comment (bfredl) [inline src/nvim/lua/converter.c:1330]: better to use `lua_tolstring` directly and avoid the alloc/free.

File: src/nvim/lua/converter.c [hunk 4]
```diff
@@ -1324,7 +1325,24 @@ void nlua_pop_keydict(lua_State *L, void *retval, FieldHashfn hashy, char **err_
...
+          int hl_id = 0;
+          if (!ERROR_SET(err) && name.size > 0) {
+            hl_id = syn_check_group(name.data, name.size);
+          }
+          api_free_string(name);
+          *(Integer *)mem = hl_id;
+        } else if (lua_type(L, -1) == LUA_TNIL) {
+          lua_pop(L, 1);
+          *(Integer *)mem = 0;
```
Comment (bfredl) [inline src/nvim/lua/converter.c:1339]: dead code. we don't enter the `lua_next()` loop at all for `nil` values.

File: src/nvim/api/private/helpers.c [hunk 5]
```diff
@@ -936,12 +936,33 @@ bool api_dict_to_keydict(void *retval, FieldHashfn hashy, Dictionary dict, Error
...
       *(Object *)mem = *value;
     } else if (field->type == kObjectTypeInteger) {
+      if (field->is_hlgroup) {
+        if (value->type != kObjectTypeNil && value->type != kObjectTypeString
+            && value->type != kObjectTypeInteger) {
+          api_set_error(err, kErrorTypeValidation, "Invalid type of '%.*s'", (int)k.size, k.data);
+          return false;
+        }
+
+        if (value->type == kObjectTypeNil) {
```
Comment (bfredl) [inline src/nvim/api/private/helpers.c:950]: after checking for nil, we can use `object_to_hl_id()` for both string, int handling, and also type error checking (`what` should be `k.data` )

**Comment by bfredl** [comment]
Nice! I pushed a small change on top to give more accurate type in `_meta/api_keysets.lua` , where we want these to be `number|string` .

**Comment by glepnir** [comment]
perfect change!

**Pull request merged**